doc/click-flatten.1
.\" -*- mode: nroff -*-
.ds V 1.5.0
.ds E " \-\- 
.if t .ds E \(em
.de Sp
.if n .sp
.if t .sp 0.4
..
.de Es
.Sp
.RS 5
.nf
..
.de Ee
.fi
.RE
.PP
..
.de Rs
.RS
.Sp
..
.de Re
.Sp
.RE
..
.de M
.BR "\\$1" "(\\$2)\\$3"
..
.de RM
.RB "\\$1" "\\$2" "(\\$3)\\$4"
..
.TH CLICK-FLATTEN 1 "21/May/2001" "Version \*V"
.SH NAME
click-flatten \- removes compound elements from a Click configuration
'
.SH SYNOPSIS
.B click-flatten
.RI \%[ options ]
.RI \%[ param = value " ...]"
.RI \%[ router\-file ]
'
.SH DESCRIPTION
The
.B click-flatten
tool flattens a Click router configuration, compiling all its compound
elements away, and writes the resulting configuration to the standard
output. It can also output information about the file.
'
.SH "OPTIONS"
'
If any filename argument is a single dash "-",
.B click-flatten
will use the standard input or output instead, as appropriate.
'
.TP 5
.BR \-c ", " \-\-classes
.PD 0
Output an alphabetical list of the element classes used by the flattened
router file, one per line, instead of the usual configuration.
'
.Sp
.TP 5
.BR \-n ", " \-\-elements
Output an alphabetical list of the element names in the flattened router
file, one per line, instead of the usual configuration.
'
.Sp
.TP 5
.BR \-d ", " \-\-declaration
Output the element declarations in the flattened router file, one per line,
instead of the usual configuration.
'
.Sp
.TP 5
.BR \-f ", " \-\-file " \fIfile"
Read the router configuration to flatten from
.IR file .
The default is the standard input.
'
.Sp
.TP 5
.BR \-e ", " \-\-expression " \fIexpr"
Use
.IR expr ,
a string in the Click language, as the router configuration to flatten.
'
.Sp
.TP
.BR \-\-config
Output flattened configuration only -- if the input was an archive, ignore
other archive elements.
'
.Sp
.TP
.BR \-\-expand\-vars
Expand global variables.  By default,
.B click-flatten
does not expand global variables into configuration strings, allowing later
operations, such as
.BR click-install ,
to override their values.
'
.Sp
.TP
.BR \-o ", " \-\-output " \fIfile"
Write the flattened router configuration to
.IR file .
The default is the standard output.
'
.Sp
.TP 5
.BI \-\-help
Print usage information and exit.
'
.Sp
.TP
.BI \-\-version
Print the version number and some quickie warranty information and exit.
'
.PD
'
.SH "SEE ALSO"
.M click 5 ,
.M click-check 1 ,
.M click-xform 1
'
.SH AUTHOR
.na
Eddie Kohler, kohler@seas.harvard.edu
.br
http://www.pdos.lcs.mit.edu/click/
'


NEWS
Click NEWS

Version 2.0.1   24.Sep.2011

* ns-3 support, thanks to Lalith Suresh and Sascha Alexander Jopen.

* Important bug fixes for user-level ToDevice thanks to Jaeyong Yoo.

* Better handling for time: Timestamp knows about steady clocks (also known
  as monotonic clocks -- clocks that never go backwards) on most platforms,
  so Timers should go off at predictable intervals even when the system
  time is set backwards.

* IPRewriter and friends: TCP supports >2GB transfers and multiple sequence
  number transitions; UDP supports different timeouts for streaming
  connections and RPC-like connections.

* Configure with --enable-stats=2 for more precise cycle counting.

* Classifier and friends (IPFilter, etc.): Correct bugs and improve
  optimization.

* Bug fixes for AddressInfo, user-level packet pool (thanks to Seiichi
  Tetsukawa).

* Support packages on patchless Linux installs.

* Thanks to Cliff Frey, and to Felipe Chaulet, Sascha Alexander Jopen,
  Christian Kreibich, Meraki, Inc., Roberto Riggio, Jim Roewe, Robert
  Sombrutzki, Lalith Suresh, Dimitris Syrivelis, Seiichi Tetsukawa, Michael
  W.S. Yee, Jaeyong Yoo, and all other contributors.


Version 2.0   16.Jul.2011

* New elements: AdjustTimestamp, BlockThread, ClickyInfo, EtherVLANEncap,
  HandlerTask, ICMPIPEncap, NullTask, SetVLANAnno, StoreUDPTimeSeqRecord,
  StripDSRHeader, StripEtherVLANHeader, TCPFragmenter, UDPIP6Encap,
  UnstripDSRHeader.

* Improved elements: ARPQuerier, AddressInfo, BandwidthRatedSplitter,
  BandwidthRatedUnqueue, CheckLength, DRRSched, FromDevice linuxmodule,
  FromDevice userlevel, FromHost userlevel, FromIPSummaryDump,
  FromSimDevice, FromUserDevice, GetIPAddress, HostEtherFilter, IPFilter
  ("ether" comparisons, faster initialization), InfiniteSource, KernelTun,
  LinkUnqueue, RandomSource, RatedSource, RatedSplitter, RatedUnqueue,
  Script, SetIPAddress, SetTimestamp, TCPIPSend, TimeSortedSched, ToDevice
  userlevel, ToUserDevice, Truncate, Unqueue2, and others.

* Element and tool bug fixes: CheckTCPHeader, CheckUDPHeader, DSRRouteTable
  DelayUnqueue, FastUDPSourceIP6, FromHost, FromUserDevice, IP6Encap,
  IPAddrRewriter, IPFilter and other classifiers, KernelFilter,
  RadiotapDecap, RadixIPLookup, SetIP6DSCP, Socket, ThreadSafeQueue,
  TimeSortedSched, ToUserDevice, UDPIPEncap, WifiDecap, click-align,
  click-undead, and others.

* Performance improvements

** Tasks: A simplified reschedule implementation and streamlined driver
   loop is both faster and less error prone (now calling fast_reschedule()
   outside of Element::run_task() is safe).

** Tasks: If a task does no work, delay re-running it.  This simple change
   improves performance on many configurations, particularly those mixing
   file descriptor or timer access and Tasks.

** Timers: A 4-heap improves performance with very large numbers of timers.

** User-level: Device elements default to libpcap even on Linux, offering
   improved performance depending on how libpcap was compiled (e.g. packet
   mmap support).  A packet memory pool suggested by Luigi Rizzo offers
   much better performance for packet allocation and freeing.

** The round-robin scheduler (--disable-stride) works again.

** New configure option --enable-bound-port-transfer slightly optimizes
   push() and pull() calls (Cliff Frey).

* Language changes

** New "require(library FILE)" allows one Click script to include another
   (Christian Kreibich).

** Many-to-one, one-to-many, and many-to-many connections cleanly express
   some connection patterns.  For example, "c :: Classifier => a1, a2, a3"
   means the same as "c :: Classifier; c [0] -> a1; c [1] -> a2; c [2] ->
   a3".

** Element groups cleanly express short detours from a connection path.
   For example, this:

	c :: Classifier(00/01);
	x -> c -> y;
	c [1] -> Paint(1) -> y;

   can be expressed like this:

	x -> c :: Classifier(00/01) => (
            [0] -> [0]; [1] -> Paint(1) -> [0]
        ) -> y;

* Multithreading

** Many performance and stability improvements to multithreaded
   configurations, particularly at userlevel.

** Lazy Task rescheduling reduces opportunities for bugs.

** Each thread gains its own sets of Timers (and, at user level, selector
   file descriptors).  An element's Timers are initially scheduled on the
   same thread as its Tasks.  This should simplify reasoning about
   Task/Timer conflicts, and can help improve performance.

* Linuxmodule

** Patchless installs are on by default.  Click installs and works without
   a patch on many versions of Linux up to 2.6.38.

* Clicky changes

** Per-router styles: ClickyInfo elements let a script specify how it
   should be drawn.

** "Run" and "Stop" buttons can execute a configuration.

* Internals

** Args: This new class and its helpers support concise, easy-to-use,
   type-safe argument parsing and are strongly preferred to cp_va_kparse in
   new code.

** TokenBucket: This new template class supports fast, flexible,
   integer-based token bucket rate management.

** Timestamps default to nanosecond precision on many platforms.

** Element::simple_action() can be used for multi-port elements.

** On some platforms, set environment variable CLICK_BACKTRACE to 1 and get
   a useful stack backtrace if Click crashes (Cliff Frey).

** Other improvements: more flexible handlers, simple_action, type traits,
   faster multi-precision arithmetic, better IPv6 address unparsing.

* Acknowledgements

** Thanks to Cliff Frey for extensive discussion, multithreaded debugging,
   802.1Q elements, TokenBucket, TCPFragmenter, bound port transfer, and
   others.

** Thanks to Joonwoo Park for Linux kernel module patches.

** Thanks to Jimmy Kjällman and Pekka Nikander for FreeBSD kernel module
   support.

** Thanks to Johan Bergs, Jesse Brown, Roman Chertov, Beyers Cronje, Jens
   De Wit, Kevin Paul Herbert, Sascha Alexander Jopen, Christian Kreibich,
   Bobby Longpocket, Ruben Merz, Derrick Pallas, Roberto Riggio, Luigi

tools/click-flatten/click-flatten.cc
/*
 * click-flatten.cc -- flatten a Click configuration (remove compounds)
 * Eddie Kohler
 *
 * Copyright (c) 2001 Mazu Networks, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, subject to the conditions
 * listed in the Click LICENSE file. These conditions include: you must
 * preserve this copyright notice, and you cannot mention the copyright
 * holders in advertising related to the Software without their permission.
 * The Software is provided WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED. This
 * notice is a summary of the Click LICENSE file; the license in that file is
 * legally binding.
 */

#include <click/config.h>
#include <click/pathvars.h>

#include <click/error.hh>
#include <click/driver.hh>
#include <click/confparse.hh>
#include "lexert.hh"
#include "routert.hh"
#include "toolutils.hh"
#include <click/clp.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

#define HELP_OPT		300
#define VERSION_OPT		301
#define CLICKPATH_OPT		302
#define ROUTER_OPT		303
#define EXPRESSION_OPT		304
#define OUTPUT_OPT		305
#define FLATTEN_OPT		306
#define CLASSES_OPT		307
#define ELEMENTS_OPT		308
#define DECLARATIONS_OPT	309
#define CONFIG_OPT		310
#define EXPAND_VARS_OPT		311

static const Clp_Option options[] = {
  { "classes", 'c', CLASSES_OPT, 0, 0 },
  { "clickpath", 'C', CLICKPATH_OPT, Clp_ValString, 0 },
  { "config", 0, CONFIG_OPT, 0, 0 },
  { "decls", 'd', DECLARATIONS_OPT, 0, 0 },
  { "declarations", 'd', DECLARATIONS_OPT, 0, 0 },
  { "elements", 'n', ELEMENTS_OPT, 0, 0 },
  { "expand-vars", 0, EXPAND_VARS_OPT, 0, Clp_Negate },
  { "expression", 'e', EXPRESSION_OPT, Clp_ValString, 0 },
  { "file", 'f', ROUTER_OPT, Clp_ValString, 0 },
  { "help", 0, HELP_OPT, 0, 0 },
  { "names", 'n', ELEMENTS_OPT, 0, 0 },
  { "output", 'o', OUTPUT_OPT, Clp_ValString, 0 },
  { "version", 'v', VERSION_OPT, 0, 0 },
};

static const char *program_name;

void
short_usage()
{
  fprintf(stderr, "Usage: %s [OPTION]... [ROUTERFILE]\n\
Try '%s --help' for more information.\n",
	  program_name, program_name);
}

void
usage()
{
  printf("\
'Click-flatten' reads a Click configuration, flattens it (removing any\n\
compound elements), and writes an equivalent configuration to the standard\n\
output.\n\
\n\
Usage: %s [OPTION]... [ROUTERFILE]\n\
\n\
Options:\n\
  -c, --classes             Output list of classes used by configuration.\n\
  -n, --names               Output list of element names in flat config.\n\
  -d, --declarations        Output list of declarations in flat config.\n\
  -f, --file FILE           Read router configuration from FILE.\n\
  -e, --expression EXPR     Use EXPR as router configuration.\n\
      --config              Output configuration only (not an archive).\n\
      --expand-vars         Expand global variables.\n\
  -o, --output FILE         Write output configuration to FILE.\n\
  -C, --clickpath PATH      Use PATH for CLICKPATH.\n\
      --help                Print this message and exit.\n\
  -v, --version             Print version number and exit.\n\
\n\
Report bugs to <click@pdos.lcs.mit.edu>.\n", program_name);
}

extern "C" {
  static int
  string_sorter(const void *va, const void *vb)
  {
    const String *a = (const String *)va, *b = (const String *)vb;
    return a->compare(*b);
  }
}

static void
output_sorted_one_per_line(Vector<String> &v, FILE *out)
{
  if (v.size())
    qsort((void *)&v[0], v.size(), sizeof(String), string_sorter);
  for (int i = 0; i < v.size(); i++)
    fprintf(out, "%s\n", v[i].c_str());
}

int
main(int argc, char **argv)
{
  click_static_initialize();
  CLICK_DEFAULT_PROVIDES;
  ErrorHandler *errh = ErrorHandler::default_handler();
  ErrorHandler *p_errh = new PrefixErrorHandler(errh, "click-flatten: ");

  // read command line arguments
  Clp_Parser *clp =
    Clp_NewParser(argc, argv, sizeof(options) / sizeof(options[0]), options);
  Clp_SetOptionChar(clp, '+', Clp_ShortNegated);
  program_name = Clp_ProgramName(clp);

  const char *router_file = 0;
  bool file_is_expr = false;
  const char *output_file = 0;
  int action = FLATTEN_OPT;
  bool expand_vars = false;

  while (1) {
    int opt = Clp_Next(clp);
    switch (opt) {

     case HELP_OPT:
      usage();
      exit(0);
      break;

     case VERSION_OPT:
      printf("click-flatten (Click) %s\n", CLICK_VERSION);
      printf("Copyright (c) 2001 Mazu Networks, Inc.\n\
Copyright (c) 2001 International Computer Science Institute\n\
Copyright (c) 2007 Regents of the University of California\n\
This is free software; see the source for copying conditions.\n\
There is NO warranty, not even for merchantability or fitness for a\n\
particular purpose.\n");
      exit(0);
      break;

     case CLICKPATH_OPT:
      set_clickpath(clp->vstr);
      break;

     case CLASSES_OPT:
     case DECLARATIONS_OPT:
     case ELEMENTS_OPT:
     case CONFIG_OPT:
      action = opt;
      break;

     case EXPAND_VARS_OPT:
      expand_vars = !clp->negated;
      break;

     case OUTPUT_OPT:
      if (output_file) {
	p_errh->error("--output file specified twice");
	goto bad_option;
      }
      output_file = clp->vstr;
      break;

     case ROUTER_OPT:
     case EXPRESSION_OPT:
     router_file:
      if (router_file) {
	p_errh->error("router configuration specified twice");
	goto bad_option;
      }
      router_file = clp->vstr;
      file_is_expr = (opt == EXPRESSION_OPT);
      break;

     case Clp_NotOption:
      if (!click_maybe_define(clp->vstr, p_errh))
	  goto router_file;
      break;

     case Clp_BadOption:
     bad_option:
      short_usage();
      exit(1);
      break;

     case Clp_Done:
      goto done;

    }
  }

 done:
  RouterT *router = read_router(router_file, file_is_expr, errh);
  if (router)
      router->flatten(errh, expand_vars);
  if (!router || errh->nerrors() > 0)
    exit(1);

  FILE *out;
  if (!output_file || strcmp(output_file, "-") == 0)
    out = stdout;
  else if (!(out = fopen(output_file, "wb"))) {
    errh->error("%s: %s", output_file, strerror(errno));
    exit(1);
  }

  switch (action) {

   case FLATTEN_OPT:
    write_router_file(router, out, errh);
    break;

   case CONFIG_OPT: {
     String s = router->configuration_string();
     ignore_result(fwrite(s.data(), 1, s.length(), out));
     break;
   }

   case CLASSES_OPT: {
     HashTable<ElementClassT *, int> m(-1);
     router->collect_types(m);
     Vector<String> classes;
     for (HashTable<ElementClassT *, int>::iterator iter = m.begin(); iter.live(); iter++)
       if (iter.key()->primitive())
	 classes.push_back(iter.key()->name());
     output_sorted_one_per_line(classes, out);
     break;
   }

   case DECLARATIONS_OPT: {
     Vector<String> decls;
     for (RouterT::iterator x = router->begin_elements(); x; x++)
       decls.push_back(x->name() + " :: " + x->type_name());
     output_sorted_one_per_line(decls, out);
     break;
   }

   case ELEMENTS_OPT: {
     Vector<String> elts;
     for (RouterT::iterator x = router->begin_elements(); x; x++)
       elts.push_back(x->name());
     output_sorted_one_per_line(elts, out);
     break;
   }

  }

  return 0;
}
